Filesystem Caching

dholms opened the pull request:
## Problem
Traversing the filesystem can take a long time especially for deeply nested & private files which require multiple back and forths to the network with decryption in between.

## Solution
- Store a "cache" in the header of dirs that contains the subtree.
- Update the cache as the subtree is updated.

This PR also includes a refactor of the way that we deal with versions:
Before we, were dealing with versions with a `normalizer` which was confusing and difficult to refactor. This has been transitioned over to inheritance.

The inheritance structure will look like
```
publicV1.0.0 -> privateV1.0.0
       |                 
       v
publicV1.0.1 -> privateV1.0.1
       |                 
       v
publicV1.1.0 -> privateV1.1.0
```
and so on. With likely breaks in the inheritance for major versions

I also added an `abstract` class `BaseTree` for most "tree" methods (such as `ls`, `cat`, `get`, etc), from which versioned trees and the "v0" `BareTree` inherit.

dholms commented on src/fs/private/tree.ts:
```diff
@@ -40,35 +28,42 @@ export class PrivateTree extends PublicTree {
-    throw new Error("This is a private node. Use PrivateNode.fromCIDEncrypted")
-  }
-
-  static async fromCIDWithKey(cid: CID, parentKey: string): Promise<PrivateTree> {
```
I'm not sure about this design choice, so I'm eager to hear feedback on it.

I ditched the separate `fromCIDWithKey` function in favor of an optional key param on `fromCID`. 
This makes writing functions that abstract over private and public trees much easier. We still throw an error if a parentKey isn't passed into `fromCID`, but we lose the type safety (and have an `undefined` which I've started trying to avoid).

Other options:
- Go back to the old way. Do runtime type checks when abstracting over private/public
- change `key?: string` to `key: Maybe<string>` which gives extra type safety (you have to intentionally pass in `null` but leads to the awkward situation where to instantiate a public tree, you have to call `PublicTree.fromCID(cid, null)`

Thoughts?

dholms commented on src/fs/v1/PrivateTree.ts:
```diff
@@ -0,0 +1,93 @@
+import semver from '../semver'
+
+
+export class PrivateTree extends PublicTree {
```
I'm not super happy with how boilerplatey a lot of this class is (also not sure happy with having 2-3 "constructors" at the bottom of each Tree/File class). But I'm not really sure a way around it. let me know if you have any ideas!

dholms commented on src/fs/bare/tree.ts:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  async createFile(content: FileContent): Promise<SimpleFile> {
+    return BareFileConstructors.create(content)
```
`BareFile` is an actual class and sometimes you want to import that for type reasons. I made it the default export. I'm not super happy with this interface tbh. I'm thinking about splitting all the constructors out into a "smart" constructor module that switches on version to select the correct constructor. That should clean up this code quite a bit, get rid of the boilerplatey stuff I talked about in my other comment, and make different types of trees better at talking to one another. Every time I've tried to do it tho I've run into circular dependency issues  :thinking:

icidasset replied:
If you use static methods like before then you could just export the class, no?

```js
class BareFile { static create() {} }
export BareFile // class

import BareFile from '...'
BareFile.create()
```

dholms replied:
Yeah I originally had them as static methods but was running into inheritance issues from private to public. Private methods take an extra param and the only way around that was to create a separate static method for private trees or to forgo type safety, make it an optional param, and throw an error if it wasn't passed

icidasset replied:
Oh ok I see, carry on then 👍

dholms replied:
Okay actually I figured out a way to do this with static methods! The difference between now and the previous time I tried to use static methods and failed is that I removed the constraint on myself that trees cannot have an internal representation of a parent key. It makes the abstraction work way better. Also I had the realization that public trees can have a parent key (meaning you have a public tree that is the child of a private tree) which makes all of the inheritance much smoother

dholms commented on src/fs/bare/tree.ts:
```diff
@@ -0,0 +1,101 @@
+    return constructors.empty()
+  }
+
+  async createTreeFromCID(cid: CID): Promise<BareTree> {
```
These instance methods are for creating children during recursive operations, and Files never have to create children

dholms commented on src/fs/bare/tree.ts:
```diff
@@ -0,0 +1,101 @@
+    return basic.putLinks(this.links, null)
+  }
+
+  async updateDirectChild(child: SimpleTree | SimpleFile, name: string): Promise<this> {
```
I introduced `SimpleTree` as an interface for a `BareTree` (ie "pretty tree"). which has slightly more limited capabilities than a `Tree` which is assumed to have a `Header`. There are a few ways around this:
- Remove the common parent (which is `BaseTree` right now). This removes a layer of abstraction and I don't think we'd need the separate interface anymore. But we would also have to duplicate a lot of the code that is shared beween `BareTree`s and `Public`/`PrivateTrees`
- hamfist `BareTree` information into a `Header` (basically add filler info, etc) so that they both fulfill the `Tree` interface. This was the approach I did before this refactor and it seemed to get pretty confusing.
- change `SimpleTree` to `Tree` and `Tree` to `HeaderTree`: this doesn't change the fact that we have two interfaces, but it might be better semantics?

If you have other thoughts I'm happy to hear them! I wasn't too happy about adding a second interface, but I didn't think the other tradeoffs were worth it

icidasset replied:
I like the idea of `Tree` and `HeaderTree`, since Tree/SimpleTree is used so much.

icidasset replied:
Maybe sticking a little README in `src/fs/` with a overview of all the types of trees could help. If you have time.

dholms commented on src/ipfs/basic.ts:
```diff
@@ -1,17 +1,23 @@
   }
   // return cid of last object (root)
-  return chunks[chunks.length - 1].cid.toString()
+  const cid = chunks[chunks.length - 1].cid.toString()
```
ipfs actually returns `CIDObj`s not just bare strings. So the type is:
```ts
type CIDObj = {
  codec: Codec
  multibaseName: MultibaseName
  string: CID
  version: number
}
```

dholms commented on src/fs/base/tree.ts:
```diff
@@ -0,0 +1,137 @@
+  async mkdir(path: string): Promise<this> {
+    const exists = await this.pathExists(path)
+    if (exists) {
+      throw new Error(`Path already exists: ${path}`)
```
Had this convo with @expede when I was first making the api, and we decided to throw an error (similar to how `mkdir` works in unix), since it gives the user more feedback as to what's happening under the hood. I was thinking it might be good to add a flag that lets you toggle that behavior

dholms commented:
@icidasset I refactored the constructors, moved them to static methods, and this actually cleaned up the implementation quite a bit. Much happier with this result than what I had before :+1:
